Login creates a new Team or adds to existing

fozcodes opened the pull request:
- If a user logs in and their Slack team _does not_ exist, it gets
  created
- If a user logs in and their Slack team _does_ exist, they are added
  with the same permissions (admin or not)

Some of this might be able to be refactored, or pulled apart. I feel like I've run in circles a couple times, so my brain is a little fried.

Suggestions/comments would be very welcome!

fozcodes committed:
Login creates a new Team or adds to existing

* If a user logs in and their Slack team _does not_ exist, it gets
created
* If a user logs in and their Slack team _does_ exist, they are added
with the same permissions (admin or not)

fozcodes committed:
Test to make sure token and uid are saved

dpdawson commented on app/models/concerns/first_or_build.rb:
```diff
@@ -0,0 +1,5 @@
+module FirstOrBuild
```
This doesn't appear to be used.

fozcodes replied:
What the? Thought I deleted that. Dammit!

dpdawson commented on app/services/authenticator.rb:
```diff
@@ -4,31 +4,65 @@ def initialize(auth_hash)
+    end
+  end
+
+  def update_membership(membership)
```
This should probably use MembershipModifier.

dpdawson commented:
As we discussed, I think all the team/user/membership creation/modification stuff should be extracted out into appropriate services/models/whatevers. Authenticator should authenticate, not do all that other stuff as well.

fozcodes committed:
Remove unused concern

fozcodes committed:
Refactor WIP

fozcodes committed:
Refactored Authenticator and service objects

* Extracted Value Objects to for Auth Hash Translator module
* Added ModelFinder module for find_or_initialize_by -> tap -> update
(FORB specific columns)
* ModelFinder module has inheritance but knows other classes - maybe should
just live in the Models instead.

fozcodes commented:
Hey Guys,

Ok, so I've refactored quite a bit here:
- Extracted Value Objects to for `AuthHashTranslator` module
  I think this is pretty useful. It cleans up all those tiny private methods that were lying around, and can be further abstracted into submodules for any additional OAuth Providers
- Added ModelFinder module for `find_or_initialize_by` -> `tap` -> `update` (FORB specific columns)
- ModelFinder module has inheritance but knows other classes - maybe should just live in the Models instead.
  I hate this name. Any suggestions? What do you call a thing that finds or initializes an object then taps it to update it further? Also, with the inheritance - these objects aren't at the leaf nodes of the object graph, so this kinda violates my rule about using inheritance.

Also wasn't sure what to call the thing that the Authenticator calls, so I went with `UserInitializer`. I really suck at naming things, so any suggestions would be welcome.

And of course, lastly, a shitload of tests are failing on Circle. They are all passing on my machine, so can someone else pull this branch and give us a tiebreaker. I'd look into it further, but my eyes feel like their about to start bleeding, so I need to go to bed.

Lemme know, thanks!

dpdawson commented:
One issue is this doesn't seem to branch from the latest develop commit. I think the duplicate migration is the real culprit, though.

dpdawson commented on app/services/auth_hash_translator/membership.rb:
```diff
@@ -0,0 +1,18 @@
+    private
+
+    def role
+      return 'admin' if auth_hash['info']['is_admin']
```
Probably should use info method.

dpdawson commented on db/migrate/20150915005416_drop_account_and_team_id.rb:
```diff
@@ -0,0 +1,6 @@
+    remove_column :teams, :account_id
+    remove_column :teams, :team_id
+  end
+end
```
Oops! We already did this in one of the last PRs.

dpdawson committed:
Remove duplicate migration.

dpdawson commented:
I rebased and fixed the migration here: https://github.com/dpdawson/ArborApp/tree/user-creates-team

I'm still processing the rest of the PR, but I think that might fix the CI failure.

dpdawson committed:
Improve naming and remove Authenticator service.

dpdawson commented:
I still don't think this is ready for primetime.

Among outstanding issues:
1) Lack of tests
2) ModelFinder service (and sub-services) don't just find, they create if the model doesn't exist. This is confusing.
3) JoinableBuilder#build_team doesn't just build a team, it uses ModelFinder, so it finds and existing team or creates a new one. It doesn't mirror JoinableBuilder#build_group, which builds a group and returns the unsaved group.
4) I think there has to be a better way to divide the concerns between JoinableBuilder, UserInitializer, and the array of ModelFinder services.

fosberyi commented on app/services/authenticator.rb:
```diff
@@ -1,47 +0,0 @@
-  def credentials
-    auth_hash['credentials']
-  end
-end
```
Good call here. This was a total middleman.

fozcodes commented:
Yeah I hear ya. I had a question there about switching to instance methods vs. class methods. As for the rest:
1. Lack of tests
- Yeah I'll get to covering this stuff
1. ModelFinder service (and sub-services) don't just find, they create if the model doesn't exist. This is confusing.
- Yeah this setup while nice and object-y, I'm regretting. I feel like these should just be methods on the respective Models. Or possibly broken into a concern module (cause the mechanism in each is very similar). Besides these subclasses making calls to the global namespace (`::{Model}`), like you said, the name sucks. And it's hard to name without being really verbose.
1. JoinableBuilder#build_team doesn't just build a team, it uses ModelFinder, so it finds and existing team or creates a new one. It doesn't mirror JoinableBuilder#build_group, which builds a group and returns the unsaved group.
- I think maybe we switch this back. I think I had the same issue with the name as you did with Model Finder (though yours is far better). It was that it was building a `Team`, but it was _creating_ a `Membership`.
  I think switching this back and moving the model finder stuff to the models will solve this.
1. I think there has to be a better way to divide the concerns between JoinableBuilder, UserInitializer, and the array of ModelFinder services.
- Yeah, I don't think that `JoinableBuilder` should be used for the login. Like you said, it's now not doing what it's name implies. The find or init'ing of the three entities during sign-in is a little more specialized than what JoinableBuilder does. I think if we move those methods to the models and have `UserInitializer` talk directly to them, it'll be much cleaner and easier to understand.

dpdawson commented:
I tend to keep services stateless unless I need to inject dependencies. They could be modules instead of classes if that makes more sense to you.

I try to limit models to persistence and associations, especially if the potential action touches multiple models (for example, creating a user, which creates a team, which creates a membership).

Maybe:
- JoinableCreator, UserCreator, etc (deal with creating a new model instance) when creation touches more than one model or is complex in some other way 
- UserFinder, TeamFinder, etc (deal with finding an existing model instance... basically a query object) if the query is complicated
- find_or_create methods could go on the respective model classes and use the finder and creator services when necessary (or this can just be handled in the controller, with the controller using the finder service, then the creator service if nothing was found)
  - some sort of Authorization service to extract that concern from the creators (could be a future refactoring)

Not sure I covered all the issues there, but let me know what you think.

fozcodes commented:
Ok, cool. It's just a matter of preference I guess. I can't make a good case against other than aesthetics.

As far as moving to the models, the methods we'd implement there would only ever touch themselves (Ha!). It'd just be that `find_or_initialize_by -> tap -> update` move in one method call. It'd be up to the object calling them to use them in concert for what we want to happen.

Lemme do it for one and see how it looks. If it's looking crap, I'll tackle it the way you're suggesting. Either way, I'll test-drive this time so we end up with tests in either case.

dpdawson commented:
It sounds what your talking about is limited to persistence, so I'm cool with it on the models. I usually just move stuff out when either a) the query gets very complex (query object) or b) the action touches multiple models or includes a number of steps (service object). Activerecord already adds a large api surface to the models, so I try to keep the methods I add on top of that simple and directly related to the same responsibility as Activerecord.

fozcodes commented:
Closing in favor of https://github.com/SandersForPresident/ArborApp/pull/29
